### PR TITLE
Fabrica SW Update Issues

### PR DESCRIPTION
**Pull request by svogt0511**
## Purpose

Vercel Preview: https://bracco-2dvm1e8if-datacite.vercel.app/ (2023/03/06)

Manual testing checklist: https://docs.google.com/document/d/18wIQyKQbiX20cOvsZclVya0CB3uGH0AUMWsuQ3fbZrs/edit?usp=sharing

Some issues surfaced as a result of the Fabrica software update.  These are the issues (and fixes) that we know of:

- (resolved) Create DOI by file upload was broken.  The DOI would be created with missing field values.  This was caused by changing the initialization of the fields in app/controllers/repositories/show/dois/upload.js.  The initialization of the fields in the doi model should have been kept to 'null' to signal lupo to resolve and save the values from what was in the xml.  Consequently, some field values would not be saved properly.
- (resolved) Modify DOI by form - the initialization in app/controllers/dois/show/modify.js needed to be reverted to its pre-update form.  Additional testing is needed to make sure all DOI field values are saved appropriately.
- (resolved) DOI 'sizes' and 'format' fields caused errors when specified in either the file upload or doi edit form.  Due to the null value initialization, they weren't being recognized as arrays which caused an error when trying to save the DOIs.  Handling them as plain arrays fixed the issues with these fields, and an additional 'set' operations in the component .js files forced notifications of change which allowed the forms to behave properly on field value changes.
- (resolved) The DOI metadata summary page formats drop down did not work.  The code box would show as empty when a metadata format was chosen via the dropdown, with errors showing in the browser console.  There were a couple of problems here.  The first was in the doi-metadata.hbs file.  The html of the code block needed to be updated.  The second issue was that subsequent switching between metadata formats using the dropdown didn't work.  The fix for this was to change the initialization of 'output' in doi-metadata.js from null/false to an empty string.
- (needs work) Some forms would freeze/do nothing if an unhandled error occurred on submit.  This is because the error information in the response was different than what was expected by the controller (see error handling in app/controllers/repositories/show/dois/upload.js).
- (resolved) Password setting seemed to be a problem for several members.  The prevailing theory is that this was a local browser cache problem which when away with a browser hard reset.

Additional items found during the latest review:
- (Cody - resolved) Prefixes tab for consortiums/consortium organizations does not work ([See discussion here.](https://github.com/datacite/bracco/pull/711#issuecomment-1446728136)).  This [commit](https://github.com/datacite/bracco/commit/c50379add1660519c72e764a67b0d50218488e5e) seems to fix it.  Apparently there is stricter syntax for referring to values.  The original syntax now refers to components.
- (Cody - resolved) Problems deleting repositories ([See discussion here.](https://github.com/datacite/bracco/pull/711#issuecomment-1446721133))  This seems to be a lupo staging environment problem, not a Fabrica problem, so we need to create an issue to investigate that.

    The Fabrica repository delete form is unresponsive using the lupo API in staging.  Looking in the browser dev console, the network request to delete the repository would show as pending for about 3 minutes, then a timeout (504) error is reported in the browser dev console.  No error reported to the user.  After a few minutes, the repository would actually show as deleted, disappearing from the repository list.  Looking at the client-api-stage logs, the DELETE network request would ultimately return 204 (success).

    I eliminated Fabrica as the problem by using postman to delete a repository in staging and observed similar behavior, a 504 - Gateway timeout, but the repository would be deleted. Using postman in the test environment to do the same request worked right away.  Then using the repository delete form from my development Fabrica pointed at the test environment worked, as well.

   As a result of investigating this, I discovered 2 additional Fabrica problems which need fixing. 

    - (resolved) First, since the form does not forward because there is an error, it needs to report the error in a flash messages box.  I added this error reporting.
    - Second, displaying the list of prefixes for the provider now shows an error mentioning in the browser console mentioning the deleted repository.  That error can be observed in the test environment display of prefixes for provider OJKL, [here](http://whirlwind.local:4200/providers/okjl/prefixes).

- (Richard - resolved) Strange error trying to update a member - only the DC member. ([See discussion here.](https://github.com/datacite/bracco/pull/711#issuecomment-1445439728)). 

    2 changes to app/controllers/providers/show/edit.js should fix this issue.

    - When an error is caught, add an error report to flashmessages so the error is reported to the user. ([Fixed here.](https://github.com/datacite/bracco/commit/4b2e6124565542aaa7720123f51dc3b39e41dc4d))
    - On provider form submission, save contacts only if you need to. ([Fixed here.](https://github.com/datacite/bracco/commit/8d7f1e0cc74997354df8a190c520b07a08ad2d1d))
  
    Some notes:

    - Works fine with my dev bracco using the test environment API.  Broken with my dev bracco using the staging environment API. Broken only for the DC member, as far as I can tell.
  
    - There are a large number of contacts for the DC consortium organization due to lack of cleanup from testing (problem 1).

    - Pressing 'submit' in the DC 'update member' form causes it to try to update all of those contacts (problem 2 - why is it updating all of the contacts?).

    - Because of the number of API requests to update DC contacts,  the AWS SQS request queue fills up (potential problem 3).  The AWS SQS queue handling lupo events is set at a maximum of 100 requests, so it starts rejecting the requests (see the screen grab, below).  At one point I had seen messages about something being 'in flight', which told me the AWS SQS request queue had something to do with it.

        - UPDATE: I found that in the submit handler for the form (submit() in app/controllers/providers/show/edit.js) there is a loop that updates every provider contact.  That explains the behavior where there is an API call to save every contact when the form is submitted.
    
         A related issue that shows up in [a flood of errors in the slack #fabrica channel](https://datacite.slack.com/archives/C6QVAK0HG/p1677280174547359) from time-to-time is probably caused by the AWS SQS request queues being overloaded in a similar manner.  This issue has been around for a while and was  not caused by the software update.  It has been happening since well before the upgrade.  I think this issue will go away when we have fixed the rest of the issues in this PR.

    - Finally, errors have occurred on submit of the 'update member' form so the form does not report back (problem 4).

        Screengrab of log entries showing the queue is full:

        91EEE0EB-9700-4647-8263-B273F5EAE3A7

Additional fixes:

- Issue #549 - where contacts disappeared from display of a provider settings tab after clicking the contacts tab and then returning to the settings tab.  This was only a display problem.  

- Updated @ember/sentry package to v6.3.1 to include fix that caused the sentry error shown in the browser console screen grab, below.  The [bug causing this error](https://github.com/getsentry/sentry-javascript/issues/3391) was fixed in [this package release](https://github.com/getsentry/sentry-javascript/releases/tag/6.3.1).

C230AD9A-4E01-4EA0-8B3C-100C0C0704A5

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)

- [ ] New feature (non-breaking change which adds functionality)

- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Reviewer, please remember our [guidelines](https://datacite.atlassian.net/wiki/spaces/TEC/pages/1168375809/Pull+Request+Guidelines):

- Be humble in the language and feedback you give, ask don't tell.
- Consider using positive language as opposed to neutral when offering feedback. This is to avoid the negative bias that can occur with neutral language appearing negative.
- Offer suggestions on how to improve code e.g. simplification or expanding clarity.
- Ensure you give reasons for the changes you are proposing.

### CODE REVIEW

**Comment by kjgarza** [comment]
Is there are vercel deployment for this? It might be important to validate that things are working there before merging.

**Comment by svogt0511** [comment]
There is a vercel branch. I have invited the group to validate in the #dev-team-chat slack channe. I never merge to main without testing and review. What you see is main merged to my fabrica-sw-update-issues branch so that we can be testing my changes on an up-to-date version of Bracco

Suzanne Vogt | Application Developer | DataCite
E: ***@***.*** | ORCID
W: datacite.org | Blog | Twitter | LinkedIn
Support Desk | Support Site | PID Forum

--- original message ---
On February 24, 2023 at 9:22 AM EST ***@***.*** wrote:



Is there are vercel deployment for this? It might be important to validate that things are working there before merging.

—

Reply to this email directly, view it on GitHub, or unsubscribe.

You are receiving this because you were assigned.Message ID: ***@***.***>
--- end of original message ---

**Comment by richardhallett** [comment]
There is a very strange error when attempting to update a member i.e. https://bracco-pbwqbkcx5-datacite.vercel.app/providers/datacite/edit
It seems to be getting a generic "server is busy response", this might be unrelated to your changes. I suggest checking staging to compare.

Otherwise I've run through various tests and as far as I can see it looks okay, another pair of eyes on this would be good still.
The code changes are fine so I've given approval but I'd wait to merge for someone else to test (and confirm the above error)

**Review by richardhallett — APPROVED** [review]

**Comment by svogt0511** [comment]
Thanks Richard. I will check the error and ask for someone else to take a look


Suzanne

--- original message ---
On February 26, 2023 at 2:09 PM EST ***@***.*** wrote:



There is a very strange error when attempting to update a member i.e. https://bracco-pbwqbkcx5-datacite.vercel.app/providers/datacite/edit

It seems to be getting a generic "server is busy response", this might be unrelated to your changes. I suggest checking staging to compare.

Otherwise I've run through various tests and as far as I can see it looks okay, another pair of eyes on this would be good still.

The code changes are fine so I've given approval but I'd wait to merge for someone else to test (and confirm the above error)

—

Reply to this email directly, view it on GitHub, or unsubscribe.

You are receiving this because you were assigned.Message ID: ***@***.***>
--- end of original message ---

**Comment by codycooperross** [comment]
When deleting a repository, the page seems to send a request to the API, but it is marked as "pending" indefinitely. The repository is not deleted and the page does not reroute. This seems to be the error: https://datacite.sentry.io/issues/3961914586/?query=is%3Aunresolved&referrer=issue-stream

**Comment by codycooperross** [comment]
I get the following error when visiting a "Prefix" tab as a Consortium Organization:

```Uncaught (in promise) TypeError: Cannot read properties of null (reading 'manager')
    at Object.evaluate (vendor-f700595424a8338ccdf9a8839ee7c316.js:sourcemap:3949:24)
    at t.evaluate (vendor-f700595424a8338ccdf9a8839ee7c316.js:sourcemap:3824:13)
    at t.evaluateSyscall (vendor-f700595424a8338ccdf9a8839ee7c316.js:sourcemap:4164:73)
    at t.evaluateInner (vendor-f700595424a8338ccdf9a8839ee7c316.js:sourcemap:4156:158)
    at t.evaluateOuter (vendor-f700595424a8338ccdf9a8839ee7c316.js:sourcemap:4156:67)
    at i.next (vendor-f700595424a8338ccdf9a8839ee7c316.js:sourcemap:4322:29)
    at e.next (vendor-f700595424a8338ccdf9a8839ee7c316.js:sourcemap:4325:93)
    at e.render (vendor-f700595424a8338ccdf9a8839ee7c316.js:sourcemap:1642:8)
    at n._renderRoots (vendor-f700595424a8338ccdf9a8839ee7c316.js:sourcemap:1671:31)
    at n._renderRootsTransaction (vendor-f700595424a8338ccdf9a8839ee7c316.js:sourcemap:1675:10)
    at n._renderRoot (vendor-f700595424a8338ccdf9a8839ee7c316.js:sourcemap:1667:50)
    at n._appendDefinition (vendor-f700595424a8338ccdf9a8839ee7c316.js:sourcemap:1663:6)
    at n.appendOutletView (vendor-f700595424a8338ccdf9a8839ee7c316.js:sourcemap:1661:6)
    at t.invoke (vendor-f700595424a8338ccdf9a8839ee7c316.js:sourcemap:4458:163)
    at t.flush (vendor-f700595424a8338ccdf9a8839ee7c316.js:sourcemap:4450:74)
    at t.flush (vendor-f700595424a8338ccdf9a8839ee7c316.js:sourcemap:4464:207)
    at r._end (vendor-f700595424a8338ccdf9a8839ee7c316.js:sourcemap:4527:9)
    at _boundAutorunEnd (vendor-f700595424a8338ccdf9a8839ee7c316.js:sourcemap:4478:574)```

**Comment by svogt0511** [comment]
> I get the following error when visiting a "Prefixes" tab as a Consortium Organization at `/providers/dc/organizations` in stage:
> 
> ```
>     Uncaught (in promise) TypeError: Cannot read properties of null (reading 'manager')
>         at Object.evaluate (vendor-f700595424a8338ccdf9a8839ee7c316.js:3949:24)
>         at t.evaluate (vendor-f700595424a8338ccdf9a8839ee7c316.js:3824:13)
>         at t.evaluateSyscall (vendor-f700595424a8338ccdf9a8839ee7c316.js:4164:73)
>         at t.evaluateInner (vendor-f700595424a8338ccdf9a8839ee7c316.js:4156:158)
>         at t.evaluateOuter (vendor-f700595424a8338ccdf9a8839ee7c316.js:4156:67)
>         at i.next (vendor-f700595424a8338ccdf9a8839ee7c316.js:4322:29)
>         at e.next (vendor-f700595424a8338ccdf9a8839ee7c316.js:4325:93)
>         at e.render (vendor-f700595424a8338ccdf9a8839ee7c316.js:1642:8)
>         at n._renderRoots (vendor-f700595424a8338ccdf9a8839ee7c316.js:1671:31)
>         at n._renderRootsTransaction (vendor-f700595424a8338ccdf9a8839ee7c316.js:1675:10)
>         at n._renderRoot (vendor-f700595424a8338ccdf9a8839ee7c316.js:1667:50)
>         at n._appendDefinition (vendor-f700595424a8338ccdf9a8839ee7c316.js:1663:6)
>         at n.appendOutletView (vendor-f700595424a8338ccdf9a8839ee7c316.js:1661:6)
>         at t.invoke (vendor-f700595424a8338ccdf9a8839ee7c316.js:4458:163)
>         at t.flush (vendor-f700595424a8338ccdf9a8839ee7c316.js:4450:74)
>         at t.flush (vendor-f700595424a8338ccdf9a8839ee7c316.js:4464:207)
>         at r._end (vendor-f700595424a8338ccdf9a8839ee7c316.js:4527:9)
>         at _boundAutorunEnd (vendor-f700595424a8338ccdf9a8839ee7c316.js:4478:574)
> ```
> 
> Seems to occur in other Consortiums as well.

I saw this error, but then the second viewing it went away.

**Comment by svogt0511** [comment]
> When deleting a repository, the page seems to send a request to the API, but it is marked as "pending" in the browser console indefinitely. The repository is not deleted and the page does not reroute. This seems to be the error: https://datacite.sentry.io/issues/3961914586/?query=is%3Aunresolved&referrer=issue-stream This does not seem to happen in production when logged in as a Consortium Organization.
> 
> _Correction:_ I eventually got confirmation emails for this repository deletion, so maybe the API is just slow on this operation.

This means an error was not caught and reported by the Fabrica interface.  The form seems to freeze on commit.  The repository is deleted.  This seems to be happening a lot.

File: app/controllers/providers/show/edit.js
Comment (jrhoads) [inline app/controllers/providers/show/edit.js:451]: Is this a typo?

Comment (jrhoads) [reply app/controllers/providers/show/edit.js:451]: `contact.eail`

**Review by codycooperross — APPROVED** [review]

**Comment by codycooperross** [comment]
Major functions look good after a manual check.

File: app/controllers/providers/show/edit.js
Comment (svogt0511) [reply app/controllers/providers/show/edit.js]: It is.  Good catch. Fixed with new vercel deployment at: https://bracco-esaolore6-datacite.vercel.app/ (2023/03/03)

**Review by jrhoads — APPROVED** [review]

**Pull request merged**
